pkg/dyninst/irgen/irgen.go

```diff
@@ -1751,13 +1753,49 @@ func populateProbeExpressions(
 	return ir.Issue{}
 }
 
+type variableToSegmentIndex map[string]int
+
+// collectAllSegmentVariables collects all variable names that are referenced by segments
+// and records the referring segments index.
+// This does assume there's no more than one segment referencing each variable
+func collectAllSegmentVariables(segments []ir.TemplateSegment) (variableToSegmentIndex, error) {
+	variableNamesUsedInTemplates := map[string]int{}
+	for i, seg := range segments {
+		jsonSeg, ok := seg.(ir.JSONSegment)
+		if !ok {
+			// Skip string segments.
+			continue
+		}
+		expr, err := exprlang.Parse(jsonSeg.JSON)
+		if err != nil {
```

Comment:
I question if this is an appropriate place for calling Parse. The alternative is in newProbe when the segments are created, but I guess adds complexity to ir.